crates/node/src/task_center.rs

```diff
@@ -0,0 +1,492 @@
+use tokio::task::JoinHandle;
+use tokio::task_local;
+use tokio_util::sync::{CancellationToken, WaitForCancellationFutureOwned};
+use tracing::{debug, error, info, instrument, warn};
+
+static SHUTDOWN_REQUESTED: AtomicBool = AtomicBool::new(false);
+static TASK_CENTER: Lazy<TaskCenter> = Lazy::new(TaskCenter::default);
+static CURRENT_EXIT_CODE: AtomicI32 = AtomicI32::new(0);
+
+const EXIT_CODE_FAILURE: i32 = 1;
+const TASK_CANCELLATION_TIMEOUT: Duration = Duration::from_secs(10);
+
+#[derive(Debug, thiserror::Error)]
+#[error("system is shutting down")]
+pub struct ShutdownError;
+
+#[derive(Default)]
+pub struct TaskCenter {
+    global_cancel_token: CancellationToken,
+}
+
+impl TaskCenter {
+    /// Used to monitor an on-going shutdown when requested
+    pub fn watch_shutdown() -> WaitForCancellationFutureOwned {
+        TASK_CENTER.global_cancel_token.clone().cancelled_owned()
+    }
+
+    /// The exit code that the process should exit with.
+    pub fn exit_code() -> i32 {
+        CURRENT_EXIT_CODE.load(Ordering::Acquire)
+    }
+
+    /// Triggers a shutdown of the system. All running tasks will be asked gracefully
+    /// to cancel within a given timeout.
+    #[instrument(level = "error", skip(exit_code))]
+    pub async fn shutdown_node(reason: &str, exit_code: i32) {
+        if SHUTDOWN_REQUESTED.load(Ordering::Acquire) {
+            // already shutting down....
+            return;
+        }
+        SHUTDOWN_REQUESTED.store(true, Ordering::Release);
+        let start = Instant::now();
+        CURRENT_EXIT_CODE.store(exit_code, Ordering::Release);
+
+        if exit_code != 0 {
+            info!("** Shutdown requested");
+        } else {
+            warn!("** Shutdown requested");
+        }
+        // todo: define a controlled shutdown by task kind as needed.
+        cancel_tasks(None, None).await;
+        // notify outer components that we have completed the shutdown.
+        TASK_CENTER.global_cancel_token.cancel();
+        info!("** Shutdown completed in {:?}", start.elapsed());
+    }
+}
+
+static TASKS: Lazy<Mutex<HashMap<TaskId, Arc<Task>>>> = Lazy::new(|| Mutex::new(HashMap::new()));
+static NEXT_TASK_ID: AtomicU64 = AtomicU64::new(0);
+
+pub struct Task {
+    /// It's nice to have a unique ID for each task.
+    id: TaskId,
+    name: &'static str,
+    // TODO: consider tracking when it was created.
+    kind: TaskKind,
+    /// cancel this token to request cancelling this task.
+    cancel: CancellationToken,
+
+    /// Tasks associated with a specific partition ID will have this set. This allows
+    /// for cancellation of tasks associated with that partition.
+    partition_id: Option<PartitionId>,
+    join_handle: Mutex<Option<JoinHandle<()>>>,
+}
+
+task_local! {
+    // This is a cancellation token which will be cancelled when a task needs to shut down.
+    static CANCEL_TOKEN: CancellationToken;
+
+    // Tasks have self-reference.
+    static CURRENT_TASK: Arc<Task>;
```

Comment:
The static is not public so that task itself won't be able to read this. Accessors like `current_task_id()` and `current_task_kind()` expose the bits we need.